### PR TITLE
BUG: Fix bug, where BooleanDtype columns are converted to Int64

### LINKED ISSUE #32287: BUG: convert_dtypes changes BooleanDtype to Int64
#### Code Sample, a copy-pastable example if possible

```python
>>> import pandas as pd
>>> df = pd.DataFrame(data=[["abc", 123, True]])
>>> print(df)
     0    1     2
0  abc  123  True
>>> print(df.dtypes)
0    object
1     int64
2      bool
dtype: object
>>> df = df.convert_dtypes()
>>> print(df)
     0    1     2
0  abc  123  True
>>> print(df.dtypes)
0     string
1      Int64
2    boolean
dtype: object
>>> df = df.convert_dtypes()
>>> print(df)
        0    1  2
0  b'abc'  123  1
>>> print(df.dtypes)
0    object
1     Int64
2     Int64
dtype: object

```
#### Problem description

Applying convert_dtypes() to a column with dtype `string` converts it to a column dtype 'object' (and the individual values from `str` type to `bytes` type).

Applying convert_dtypes() to a column with dtype `boolean` converts it to a column dtype 'Int64' (and the individual values from `bool` type to `int` type).

#### Expected Output

convert_dtypes() should keep StringDtype columns as StringDtype and BooleanDtype columns as BooleanDtype.

#### Output of ``pd.show_versions()``

<details>

pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_Malaysia.1252

pandas           : 1.0.1
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.2.3
setuptools       : 41.2.0
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None

</details>

### PR DESCRIPTION
**Pull request by AnnaDaglis**
- [x] closes #32287
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -1078,8 +1080,9 @@ def convert_dtypes(
                 inferred_dtype = input_array.dtype
 
         if convert_boolean:
-            if is_bool_dtype(input_array.dtype) and not is_extension_array_dtype(
-                input_array.dtype
+            if is_bool_dtype(input_array.dtype) and (
+                (not is_extension_array_dtype(input_array.dtype))
+                or (input_array.dtype == pd.BooleanDtype())
```
Comment (WillAyd) [inline pandas/core/dtypes/cast.py]: Can this just be simplified to `is_bool_dtype`?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/dtypes/cast.py]: I think the issue is actually earlier. The condition at https://github.com/pandas-dev/pandas/pull/32490/files#diff-1df6c746d2a342e60cce09f774b566d7R1073-R1076 isn't wanted here.

Should this entire check be inside a big `if not is_extension_array_dtype(input_array.dtype)`? There aren't any cases where we want to convert from extension type to extension type, right?

Comment (TomAugspurger) [reply pandas/core/dtypes/cast.py]: And by "entire check" I mean immediately following the `if convert_string or convert_integer or convert_boolean` at https://github.com/pandas-dev/pandas/pull/32490/files#diff-1df6c746d2a342e60cce09f774b566d7R1054

File: pandas/tests/series/methods/test_convert_dtypes.py [hunk 2]
```diff
@@ -279,3 +279,16 @@ def test_convert_string_dtype(self):
         )
         result = df.convert_dtypes()
         tm.assert_frame_equal(df, result)
+
+    def test_convert_bool_dtype(self):
+        # GH32287
+        df = pd.DataFrame([["abc", 123, True]])
```
Comment (TomAugspurger) [inline pandas/tests/series/methods/test_convert_dtypes.py]: A bit simpler test

```suggestion
        df = pd.DataFrame({"A": pd.array([True])})
        tm.assert_frame_equal(df, df.convert_dtypes())
```

**Comment by TomAugspurger** [comment]
@AnnaDaglis do you think you'll have time to update this soon? We're hoping to get 1.0.2 out in the next couple days. If not, no problem, we'll just move it to the next release.

**Comment by jorisvandenbossche** [comment]
> Should this entire check be inside a big if not is_extension_array_dtype(input_array.dtype)? There aren't any cases where we want to convert from extension type to extension type, right?

Yes, right now, if the dtype is already an extension dtype, we don't want to change it (in the future, if there is eg a new nullable timestamp dtype, we might want to do this, but that is not yet relevant right now)

**Comment by TomAugspurger** [comment]
Ha @AnnaDaglis I was just pushing the same changes here. Looks like we ran into the same issue with the Series.astype failing to copy.

Mind if I push some commits here? I think I tracked it down.

**Comment by TomAugspurger** [comment]
I might end up just force pushing to your branch if that's OK. It'll mess some things up but will be easiest I think.

**Comment by AnnaDaglis** [comment]
@TomAugspurger No problem!

**Comment by TomAugspurger** [comment]
OK thanks! If you want to sync up, you can `git fetch origin` and then `git reset --hard origin/fix_convert_dtypes_bool` when you have that branch checked out locally. Hopefully we're good to go.

@jbrockmendel on master `Series.astype()` wasn't always copying for datetime64 dtypes. Those changes are in 934147b492d265cfd798c5288821f665c82a7a7c.

**Comment by jbrockmendel** [comment]
nice catch

**Review by TomAugspurger — APPROVED** [review]
Tagging for 1.0.2 since this is NA related.

**Comment by jreback** [comment]
thanks @AnnaDaglis very nice!

**Comment by jreback** [comment]
@meeseksdev backport 1.0.x

**Comment by jreback** [comment]
@meeseeksdev backport 1.0.x

**Comment by TomAugspurger** [comment]
Huh, this seemingly wasn't backported, but the release notes are there... I'll do something manually.

**Pull request merged**